### ISSUE TITLE
Startup Time

### ISSUE THREAD

**Issue by Sh3Rm4n** [nvim-treesitter/nvim-treesitter#212]
TLDR: Maybe lazy-load `utils.setup_command` functions? 

I did some `--startuptime` analysis on my configuration today and noticed, that
`nvim-treesitter` does add a fair amount delay to the startup.

I'm using a rather slow system:

- Windows 10 1909 and nvim inside WSL

After playing around a bit, I noticed that the `utils.setup_command` function
have a great effect on startup.

Commenting out these lines:

https://github.com/nvim-treesitter/nvim-treesitter/blob/954b8c718f09e3cb3c699d91b01b283bd48146a0/lua/nvim-treesitter.lua#L11-L12

saves `~250ms` from `~500ms` before.

**To Reproduce**

Steps to reproduce my very rough measurements:

1. `nvim file-with-no-treesitter-support.rst --startuptime time.txt`
2. Looking at the relevant part of `time.txt`:
```log
1604.056  555.734  555.734: sourcing /home/viol/.local/share/nvim/plugged/nvim-treesitter/plugin/nvim-treesitter.vim
```
3. Commenting out the mentioned lines from above
4. Repeating step 1
5. Readout new times from `time.txt`
```log
1191.265  260.503  260.503: sourcing /home/viol/.local/share/nvim/plugged/nvim-treesitter/plugin/nvim-treesitter.vim
```

**Expected behavior**

I may have been not really careful and other parts could have also played a part
of the reduced startup time. I just wanted to bring attention to the startup
time in general. Maybe there are ways to optimize it in the future? 
If I understand in correctly the `utils.setup_command` functions are not really
needed for the overall functionality and could also be loaded lazily?

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `git` executable found.
  - OK: `cc` executable found.

## html parser healthcheck
  - OK: html parser found.
  - OK: `highlights.scm` found.
  - WARNING: No `locals.scm` query found for html
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for html
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## typescript parser healthcheck
  - OK: typescript parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for typescript
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## markdown parser healthcheck
  - OK: markdown parser found.
  - WARNING: No `highlights.scm` query found for markdown
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for markdown
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for markdown
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## regex parser healthcheck
  - OK: regex parser found.
  - OK: `highlights.scm` found.
  - WARNING: No `locals.scm` query found for regex
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for regex
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## c parser healthcheck
  - OK: c parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - OK: `textobjects.scm` found.

## nix parser healthcheck
  - OK: nix parser found.
  - WARNING: No `highlights.scm` query found for nix
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for nix
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for nix
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## swift parser healthcheck
  - OK: swift parser found.
  - WARNING: No `highlights.scm` query found for swift
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for swift
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for swift
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## java parser healthcheck
  - OK: java parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for java
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## python parser healthcheck
  - OK: python parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - OK: `textobjects.scm` found.

## yaml parser healthcheck
  - OK: yaml parser found.
  - WARNING: No `highlights.scm` query found for yaml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for yaml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for yaml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## elm parser healthcheck
  - OK: elm parser found.
  - WARNING: No `highlights.scm` query found for elm
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for elm
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for elm
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## vue parser healthcheck
  - OK: vue parser found.
  - WARNING: No `highlights.scm` query found for vue
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for vue
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for vue
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## cpp parser healthcheck
  - OK: cpp parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - OK: `textobjects.scm` found.

## toml parser healthcheck
  - OK: toml parser found.
  - WARNING: No `highlights.scm` query found for toml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for toml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for toml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## lua parser healthcheck
  - OK: lua parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for lua
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## ruby parser healthcheck
  - OK: ruby parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for ruby
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## ocaml parser healthcheck
  - OK: ocaml parser found.
  - WARNING: No `highlights.scm` query found for ocaml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for ocaml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for ocaml
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## go parser healthcheck
  - OK: go parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - OK: `textobjects.scm` found.

## scala parser healthcheck
  - OK: scala parser found.
  - WARNING: No `highlights.scm` query found for scala
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for scala
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for scala
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## haskell parser healthcheck
  - OK: haskell parser found.
  - WARNING: No `highlights.scm` query found for haskell
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for haskell
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for haskell
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## rust parser healthcheck
  - OK: rust parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - OK: `textobjects.scm` found.

## json parser healthcheck
  - OK: json parser found.
  - OK: `highlights.scm` found.
  - WARNING: No `locals.scm` query found for json
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for json
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## jsdoc parser healthcheck
  - OK: jsdoc parser found.
  - WARNING: No `highlights.scm` query found for jsdoc
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for jsdoc
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for jsdoc
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## javascript parser healthcheck
  - OK: javascript parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for javascript
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## css parser healthcheck
  - OK: css parser found.
  - OK: `highlights.scm` found.
  - WARNING: No `locals.scm` query found for css
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for css
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## julia parser healthcheck
  - OK: julia parser found.
  - WARNING: No `highlights.scm` query found for julia
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for julia
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for julia
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## php parser healthcheck
  - OK: php parser found.
  - WARNING: No `highlights.scm` query found for php
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `locals.scm` query found for php
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `textobjects.scm` query found for php
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## c_sharp parser healthcheck
  - OK: c_sharp parser found.
  - OK: `highlights.scm` found.
  - WARNING: No `locals.scm` query found for c_sharp
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - OK: `textobjects.scm` found.

## bash parser healthcheck
  - OK: bash parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for bash
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

## tsx parser healthcheck
  - OK: tsx parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for tsx
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

</code>
</details>

**Output of `nvim --version`**
<details>

```

NVIM v0.5.0-610-g326b87feb
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/sbin/cc -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/viol/.cache/yay/neovim-git/src/build/config -I/home/viol/.cache/yay/neovim-git/src/neovim-git/src -I/usr/include -I/home/viol/.cache/yay/neovim-git/src/build/src/nvim/auto -I/home/viol/.cache/yay/neovim-git/src/build/include
Compiled by viol@DE2WXL-520036

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

</details>

**Comment by Sh3Rm4n** [comment]
Well maybe `bug` isn't really the right label. It fits `enhancement` better IMO.

**Comment by Sh3Rm4n** [comment]
Well after some more investigation, I've found out, that my measurement is not really correct and it has more to do with the recursive discovery of modules (via `define_modules`). I think this issue, as currently described, is not really helpful.
